---
title: "Create Illumina Sample Sheet and metadata"
author: "Ramon Gallego"
date: "11/8/2019"
output: html_document
params: 
  Assay: "Nextera XT"
  Index_Adapters: "Nextera XT Index Kit (96 Indexes 384 Samples)"
  Cycles_per_pairend: 301
  Set: 1
  input.metadata: ../data_sub/metadata.csv
  date: !r Sys.Date()
  output_dir: ../data_sub/
---
## How to use this

To run this Rmarkdown, open it in Rstudio and from the knit drop-down menu, choose Knit with parameters
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library (here)
#TODO: Change input to include i7 and i5 primers - and use that info
```

## What does this do

It creates the Illumina Sample sheet and the `metadata.csv` file that you need for running the pipeline. You need to double check the output - weird things happen all the time.

## Required input

You need an input file with the following columns:

  - Sample name
  - Well in the 96-well plate used for the Nextera reaction. 
  - PrimerF, 
  - PrimerR,
  - i7 Adapter (name) used to identify samples. Format is N7XX
  - i5 Adapter (name) used to identify samples. Format is N5XX


## Loading the metadata you want to sequence

```{r load metadata}

init.metadata <- read_csv(params$input.metadata)

init.metadata

```

### Check that the metadata has all needed fields

```{r Check metadata}

init.metadata %>% 
  rownames_to_column("Sample_number") %>% 
  mutate_all(as.character) %>% 
  pivot_longer(-Sample_number, names_to = "Variable", values_to = "Value") %>% 
  summarise(Sample_name = case_when(sum(str_detect(Variable, "Sample")) > 0 ~ "Present",
                                 TRUE                                    ~ "Absent"),
         Well        = case_when(sum(str_detect(Variable, "Well")) > 0 ~ "Present",
                                 TRUE                                    ~ "Absent"),
         PrimerF    = case_when(sum(str_detect(Variable, "PrimerF")) > 0 ~ "Present",
                                 TRUE                                    ~ "Absent"),
         PrimerR    = case_when(sum(str_detect(Variable, "PrimerR")) > 0 ~ "Present",
                                 TRUE                                    ~ "Absent"),
         
         i5_Adapter  = case_when(sum(str_detect(Variable, "i5_Adapter")) > 0 ~ "Present",
                                 TRUE                                    ~ "Absent"),
         i7_Adapter  = case_when(sum(str_detect(Variable, "i7_Adapter")) > 0 ~ "Present",
                                 TRUE                                    ~ "Absent")) -> Checks
  


if(sum(str_detect(Checks,"Absent")) > 0){ knitr::knit_exit(append = "## ERROR: Initial metadata is missing some of the key column names: 

                                                           - Sample_name
                                                           - Well
                                                           - PrimerF
                                                           - PrimerR
                                                           - Plate
                                                           - i5_Adapter
                                                           - i7_Adapter
                                                           
                                                           ### Change the column names / add the infromation to your csv file")}

```


## Load the Illumina SampleSheet template



```{r Loading SampleSheet}
template.sample.sheet <- read_lines(here("data_sub","SampleSheetLibPrep.csv"))
```

Locate all the lines in which the new parameters are going to be written

```{r locate terms}


date.row  <- str_which(template.sample.sheet, "^\\Date")
Assay.row <- str_which(template.sample.sheet, "^Assay")
Index.row <- str_which(template.sample.sheet, "^Index Adapters")
Reads.row <- str_which(template.sample.sheet, "^\\[Reads]")
data.start <- str_which(template.sample.sheet, "^\\[Data]")
data.header <- str_which(template.sample.sheet, "^Sample_ID")

```

## Load the Illumina adapters

```{r load nextera}

Nextera_Adapters_i7 <- read_csv(here("data_sub","Nextera_adapters_i7.csv"))

Nextera_Adapters_i5 <- read_csv(here("data_sub","Nextera_adapters_i5.csv")) %>%
  mutate(I5_Index_ID = str_replace(I5_Index_ID, "S", "N"))

Nextera_Adapters_i7 %>%
  mutate (Set = rep(c(1,2,3), each = 12)[1:26]) -> Nextera_Adapters_i7

Nextera_Adapters_i5 %>%
  mutate (Set = rep(c(1,2,3), each = 8)[1:18]) -> Nextera_Adapters_i5

```

## Merge metadata with illumina adapters

```{r merging}
# Nextera_Adapters_i7 %>% 
#   filter (Set == params$Set) %>% 
#   select(-Position, -`Bases in Adapter`, -Set) %>% 
#   right_join(init.metadata, by = c("Column")) %>% 
#   left_join(Nextera_Adapters_i5 %>% 
#               filter(Set == params$Set) %>% 
#               select(-Position, -`Bases in Adapter`)) %>% 
#   select(Sample, everything()) -> metadata
init.metadata %>% 
  left_join(Nextera_Adapters_i7 %>% 
              select(I7_Index_ID,index = `Bases in Adapter`),
             by = c( "i7_Adapter"= "I7_Index_ID" ) ) %>% 
  left_join(Nextera_Adapters_i5 %>% 
              select(I5_Index_ID,index2 = `Bases in Adapter`),
            by = c("i5_Adapter" = "I5_Index_ID")) -> metadata

```

## Fill Illumina SampleSheet

First with the parameters not linked with each sample
### If there are two or more sets of PCR primers, we need to limit the sample sheet to the unique combinations of sample and barcodes, and split those in the pipeline

```{r}
template.sample.sheet[date.row] <- paste0("Date,",Sys.Date())

template.sample.sheet[Assay.row] <- paste0("Assay," ,params$Assay)
template.sample.sheet[Index.row] <- paste0("Index Adapters,",params$Index_Adapters)
template.sample.sheet[Reads.row +2 ] <- template.sample.sheet[Reads.row + 4] <- params$Cycles_per_pairend
```

Then with the rest of the dataset - collapsing first by primers

```{r collapse by primer}

metadata %>% 
  group_by(i5_Adapter,i7_Adapter ) %>% 
  tally() -> occurrence.of.combos

if(max(occurrence.of.combos$n > 1 )){
  print("There are combinations of barcodes assigned to more than 1 sample - Please Check they have different PCR primers so you can separate them after")
  metadata %>% 
  group_by(i5_Adapter,i7_Adapter ) %>% 
    slice(1) -> metadata.for.sample.sheet
  print ("Keeping the first occurence of each combo for your samplesheet")}else{metadata.for.sample.sheet <- metadata}
```


```{r}
sample.data <- tibble (Sample_Name = metadata.for.sample.sheet$Sample_name,
                      Sample_Plate = params$Set,
                      Sample_Well = metadata.for.sample.sheet$Well,
                      I7_Index_ID = metadata.for.sample.sheet$i7_Adapter,
                      index       = metadata.for.sample.sheet$index,
                      I5_Index_ID = metadata.for.sample.sheet$i5_Adapter,
                      index2      = metadata.for.sample.sheet$index2,
                      Sample_Project ="", 
                      Description = "") %>% 
                rownames_to_column("Sample_ID")


nsamples <- nrow(sample.data)
template.sample.sheet <- template.sample.sheet[1:data.header]
for (i in 1:nsamples){
  
  template.sample.sheet[i+data.header] <- paste(sample.data[i,], collapse = ",") 
  
  }
template.sample.sheet

write_lines(template.sample.sheet, 
            paste0(params$output_dir, "SampleSheet_",Sys.Date(),".csv"),
            sep = "\r\n")
```

## Fill in metadata

We are going to assume that the Illumina always returns files with the format `Sample_name` `_L001_R` [1-2] '_001.fastq`. So let's fill the metadata accordingly

```{r}
metadata %>%
  mutate(file1 = paste0(Sample_name, "_L001_R1_001.fastq"),
         file2 = paste0(Sample_name, "_L001_R2_001.fastq")) %>%
  group_by(PrimerF, PrimerR) %>% 
  nest() %>% 
  rownames_to_column("Locus") %>% 
  unnest(data) %>% 
  select(Sample_name,
         file1,
         file2,
         i7_Adapter,
         i5_Adapter,
         Well,
         PrimerF,
         PrimerR, Locus) -> metadata


write_csv(metadata, paste0(params$output_dir, "metadata_",Sys.Date(),".csv"))
```

